File: 01.Python/1186.py
from typing import List

class Solution:
    def maximumSum(self, arr: List[int]) -> int:
        noDelSum = arr[0]
        oneDelSum = -10_000
        result = arr[0]

        for i in range(1, len(arr)):
            prevNoDel = noDelSum

            noDelSum = max(arr[i], arr[i] + noDelSum)
            oneDelSum = max(prevNoDel, arr[i] + oneDelSum)

            result = max(result, noDelSum, oneDelSum)
        
        return result

sol = Solution()

arr1 = [1, -2, 0, 3]
result1 = sol.maximumSum(arr1)

arr2 = [1, -2, -2, 3]
result2 = sol.maximumSum(arr2)

arr3 = [-1, -1, -1, -1]
result3 = sol.maximumSum(arr3)

print(result1)
print(result2)
print(result3)

File: 04.C/1186.c
#include <stdio.h>

int max(int a, int b) {
    return (a > b) ? a : b;
}

int maximumSum(int* arr, int arrSize) {
    int noDelSum = arr[0];
    int oneDelSum = -10000;
    int result = arr[0];

    for (int i = 1; i < arrSize; i++) {
        int prevNoDel = noDelSum;

        noDelSum = max(
            arr[i],
            noDelSum + arr[i]
        );

        oneDelSum = max(
            prevNoDel,
            arr[i] + oneDelSum
        );

        result = max(
            result,
            max(noDelSum, oneDelSum)
        );
    }

    return result;
}

int main() {
    int arr1[] = {1, -2, 0, 3};
    int arr1Size = sizeof(arr1) / sizeof(arr1[0]);
    int result1 = maximumSum(arr1, arr1Size);
    printf("%d\n", result1);

    int arr2[] = {1, -2, -2, 3};
    int arr2Size = sizeof(arr2) / sizeof(arr2[0]);
    int result2 = maximumSum(arr2, arr2Size);
    printf("%d\n", result2);

    int arr3[] = {-1, -1, -1, -1};
    int arr3Size = sizeof(arr3) / sizeof(arr3[0]);
    int result3 = maximumSum(arr3, arr3Size);
    printf("%d\n", result3);

    return 0;
}